### PR TITLE
Add kafka 2.3.1 support

### PR DESCRIPTION
**Pull request by tomncooper**
### Type of change

- Enhancement

### Description

- Adds support for Kafka 2.3.1 
- Refactors Cluster Config test to remove any hardcoded versions strings

### Checklist

- [x] Make sure all tests pass
- [x] Update documentation
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Update CHANGELOG.md

### CODE REVIEW

**Comment by scholzj** [comment]
@strimzi-ci run tests

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/KafkaVersionTestUtils.java [hunk 1]
```diff
@@ -14,25 +14,28 @@
...
     public static final String KAFKA_IMAGE_STR = "strimzi/kafka:latest-kafka-";
     public static final String KAFKA_CONNECT_IMAGE_STR = "strimzi/kafka-connect:latest-kafka-";
     public static final String KAFKA_CONNECT_S2I_IMAGE_STR = "strimzi/kafka-connect-s2i:latest-kafka-";
     public static final String KAFKA_MIRROR_MAKER_IMAGE_STR = "strimzi/kafka-mirror-maker:latest-kafka-";
 
-    public static final String LATEST_KAFKA_VERSION = "2.3.1";
+    public static final String LATEST_KAFKA_VERSION = "2.3.0";
     public static final String LATEST_FORMAT_VERSION = "2.3";
     public static final String LATEST_PROTOCOL_VERSION = "2.3";
+    public static final String LATEST_ZOOKEEPER_VERSION = "3.5.4";
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/KafkaVersionTestUtils.java]: How do these changes fit with this PR?

Comment (tomncooper) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/KafkaVersionTestUtils.java]: This is part of the ZK upgrade PR. I originally had the changes to this class in there to help with the 2.4.0 testing and patched them into this PR as they are relevant to testing version changes. I thought I had removed all the ZK upgrade specific sections but I missed this one. I will fix this and double check that there aren't any more.

**Comment by ppatierno** [comment]
@strimzi-ci run tests

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/KafkaVersionTestUtils.java [hunk 2]
```diff
@@ -14,25 +14,28 @@
...
     public static final String KAFKA_VERSION_YAML_TEMPLATE =
-                    "- version: %s\n  format: %s\n  protocol: %s\n  checksum: %s\n  third-party-libs: %s\n  default: %b\n";
+                    "- version: %s\n  format: %s\n  protocol: %s\n  zookeeper: %s\n  checksum: %s\n  third-party-libs: %s\n  default: %b\n";
     public static final String KAFKA_IMAGE_STR = "strimzi/kafka:latest-kafka-";
     public static final String KAFKA_CONNECT_IMAGE_STR = "strimzi/kafka-connect:latest-kafka-";
     public static final String KAFKA_CONNECT_S2I_IMAGE_STR = "strimzi/kafka-connect-s2i:latest-kafka-";
     public static final String KAFKA_MIRROR_MAKER_IMAGE_STR = "strimzi/kafka-mirror-maker:latest-kafka-";
 
-    public static final String LATEST_KAFKA_VERSION = "2.3.1";
+    public static final String LATEST_KAFKA_VERSION = "2.3.0";
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/KafkaVersionTestUtils.java]: Why are you changing this back to 2.3.0? Just curious, maybe you know more about what it does :-o

**Comment by ppatierno** [comment]
@strimzi-ci run tests

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/KafkaVersionTestUtils.java (hunk 2 above)
Comment (tomncooper) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/KafkaVersionTestUtils.java]: TBH I am not sure I had a very strong reason for changing this back, but the actual values in this class are not really relevant. The idea of this util class is to remove the need for any hardcoded version numbers in the tests.

File: api/src/main/java/io/strimzi/api/kafka/model/JvmOptions.java [hunk 3]
```diff
@@ -32,7 +32,7 @@
     private String xmx;
     private String xms;
     private Boolean server;
-    private boolean gcLoggingEnabled = false;
+    private boolean gcLoggingEnabled = true;
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/JvmOptions.java]: Why was this changed?

File: documentation/modules/snip-kafka-versions.adoc [hunk 4]
```diff
@@ -6,6 +6,7 @@
 [options="header"]
 |=================
 |Kafka version |Interbroker protocol version |Log message format version| ZooKeeper version
-| 2.2.1 | 2.2 | 2.2 | 3.4.13
-| 2.3.0 | 2.3 | 2.3 | 3.4.13
+| 2.2.1 | 2.2 | 2.2 | 3.4.14
```
Comment (tombentley) [inline documentation/modules/snip-kafka-versions.adoc]: Kafka 2.2.1 shipped with ZK 3.4.13.

File: documentation/snip-kafka-versions.sh [hunk 5]
```diff
@@ -25,7 +25,7 @@ cat <<EOF
 EOF
 for i in "${!versions[@]}"
 do 
-    echo "| ${versions[$i]} | ${protocols[$i]} | ${formats[$i]} | 3.4.13"
+    echo "| ${versions[$i]} | ${protocols[$i]} | ${formats[$i]} | 3.4.14"
```
Comment (tombentley) [inline documentation/snip-kafka-versions.sh]: As noted on the file this script generates, it's not as simple as this.

File: api/src/main/java/io/strimzi/api/kafka/model/JvmOptions.java (hunk 3 above)
Comment (tomncooper) [reply api/src/main/java/io/strimzi/api/kafka/model/JvmOptions.java]: That was in @scholzj's original PR.

File: documentation/snip-kafka-versions.sh (hunk 5 above)
Comment (tomncooper) [reply documentation/snip-kafka-versions.sh]: Yeah 2.2.1 has 3.4.13 and 2.3.0 has 3.4.14. This issue is solved in the ZK upgrade PR as the versions are in the `kafka-versions.yaml` and so this script parses them correctly. We could put an if statement in, but that would be messy. Is it worth changing this? Or adding a note?

File: api/src/main/java/io/strimzi/api/kafka/model/JvmOptions.java (hunk 3 above)
Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/JvmOptions.java]: @scholzj I was going to ask the same, why this?

Comment (tomncooper) [reply api/src/main/java/io/strimzi/api/kafka/model/JvmOptions.java]: I am gonna revert this unless @scholzj has a strong objection?

File: documentation/snip-kafka-versions.sh (hunk 5 above)
Comment (tomncooper) [reply documentation/snip-kafka-versions.sh]: I have patched in the zookeeper version features from my zookeeper upgrade branch so this should be solved now.
